z3/src/datatype_builder.rs

```diff
@@ -1,117 +1,179 @@
+    let num = ds.len();
+    assert!(num > 0, "At least one DatatypeBuilder must be specified");
+
+    let ctx: &'ctx Context = ds[0].ctx;
+    let mut names: Vec<Z3_symbol> = Vec::with_capacity(num);
+
+    let out: Vec<Z3_sort> = Vec::with_capacity(num);
+    // Prevent running `out's destructor so we are in complete control
+    // of the allocation.
+    let mut out = mem::ManuallyDrop::new(out);
+    let mut clists: Vec<Z3_constructor_list> = Vec::with_capacity(num);
+
+    for d in ds.iter() {
+        names.push(d.name.as_z3_symbol(ctx));
+        let num_cs = d.constructors.len();
+        let mut cs: Vec<Z3_constructor> = Vec::with_capacity(num_cs);
+
+        for (cname, fs) in &d.constructors {
+            let mut rname: String = "is-".to_string();
+            rname.push_str(cname);
+
+            let cname_symbol: Z3_symbol = Symbol::String(cname.clone()).as_z3_symbol(ctx);
+            let rname_symbol: Z3_symbol = Symbol::String(rname).as_z3_symbol(ctx);
+
+            let num_fs = fs.len();
+            let mut field_names: Vec<Z3_symbol> = Vec::with_capacity(num_fs);
+            let mut field_sorts: Vec<Z3_sort> = Vec::with_capacity(num_fs);
+            let mut sort_refs: Vec<::std::os::raw::c_uint> = Vec::with_capacity(num_fs);
+
+            for (fname, accessor) in fs {
+                field_names.push(Symbol::String(fname.clone()).as_z3_symbol(ctx));
+                match accessor {
+                    DatatypeAccessor::Datatype(dtype_name) => {
+                        field_sorts.push(null_mut());
+
+                        let matching_names: Vec<_> = ds
+                            .iter()
+                            .enumerate()
+                            .filter(|&(i, x)| &x.name == dtype_name)
+                            .collect();
+
+                        assert_eq!(
+                            1,
+                            matching_names.len(),
+                            "One and only one occurrence of each datatype is expected."
+                        );
+
+                        let (sort_ref, _) = matching_names[0];
+                        sort_refs.push(sort_ref as u32);
+                    }
+                    DatatypeAccessor::Sort(sort) => {
+                        field_sorts.push(sort.z3_sort);
+                        sort_refs.push(0);
+                    }
+                }
             }
-        };
-
-        // create independent fields
-        let (ctx, variants) = (self.ctx, self.variants);
 
-        let variants = variants
-            .into_iter()
-            .map(|(num_fields, constructor)| {
-                let mut constructor_func: Z3_func_decl = null_mut();
-                let mut tester: Z3_func_decl = null_mut();
-                let mut accessors: Vec<Z3_func_decl> = Vec::new();
-                accessors.resize(num_fields, null_mut());
+            let constructor = unsafe {
+                Z3_mk_constructor(
```

Comment:
We should `let _guard = Z3_MUTEX.lock();` inside the scope of each raw `z3_sys` function call.

Be careful to introduce small block scopes, if necessary, to avoid holding the lock too long or across other higher level library calls that internally try to take the lock and trigger a dead lock (which is annoying to debug).

Reply:
Done.